cpp/src/arrow/flight/client.cc

```diff
@@ -259,6 +111,79 @@ class FlightStreamReader : public RecordBatchReader {
   std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream_;
 };
 
+class FlightClient;
+
+/// \brief A RecordBatchWriter implementation that writes to a Flight
+/// DoPut stream.
+class FlightPutWriter::FlightPutWriterImpl : public ipc::RecordBatchWriter {
+ public:
+  explicit FlightPutWriterImpl(std::unique_ptr<ClientRpc> rpc,
+                               const FlightDescriptor& descriptor,
+                               const std::shared_ptr<Schema>& schema,
+                               MemoryPool* pool = default_memory_pool())
+      : rpc_(std::move(rpc)), descriptor_(descriptor), schema_(schema), pool_(pool) {}
+
+  Status WriteRecordBatch(const RecordBatch& batch, bool allow_64bit = false) override {
+    IpcPayload payload;
+    RETURN_NOT_OK(ipc::internal::GetRecordBatchPayload(batch, pool_, &payload));
+    auto custom_writer = reinterpret_cast<grpc::ClientWriter<IpcPayload>*>(writer_.get());
+    // Explicitly specify the override to invoke - otherwise compiler
+    // may invoke through vtable (not updated by reinterpret_cast)
+    if (!custom_writer->grpc::ClientWriter<IpcPayload>::Write(payload,
+                                                              grpc::WriteOptions())) {
+      // Stream ended?
+      return Status::UnknownError("Could not write record batch to stream");
```

Comment:
Apparently it might be possible to get `ClientContext::debug_error_string` (https://grpc.io/grpc/cpp/classgrpc_1_1_client_context.html#aa31ee946b2f51e07e2f62f7866f42f20) to get a better error message?

In any case, this shouldn't be `UnknownError` but `IOError`.